onnxruntime/test/ir/graph_test.cc

```diff
@@ -1527,6 +1588,56 @@ TEST_F(GraphTest, AddRemoveInitializerHandling) {
                                  << num_initializers << " remain.";
 }
 
+TEST_F(GraphTest, SparseInitializerHandling) {
+  const char* const input_initializer_name = "x";
+  Model model("SparseInitializerHandling", false, *logger_);
+  std::string s1;
+  // Create model proto with sparse initializer
+  {
+    auto model_proto = model.ToProto();
+    auto* m_graph = model_proto.mutable_graph();
+    ConstructASimpleAddGraph(*m_graph, nullptr);
+    auto* m_sparse_initializer = m_graph->add_sparse_initializer();
+    ConstructSparseTensor(input_initializer_name, *m_sparse_initializer);
+    model_proto.SerializeToString(&s1);
+  }
+
+  ModelProto model_proto_sparse;
+  const bool result = model_proto_sparse.ParseFromString(s1);
+  ASSERT_TRUE(result) << "Failed to load model from serialized protobuf";
+  {
+    auto& graph_proto = model_proto_sparse.graph();
+    ASSERT_EQ(graph_proto.initializer_size(), 0);
+    ASSERT_EQ(graph_proto.sparse_initializer_size(), 1);
+  }
+
+  std::shared_ptr<onnxruntime::Model> p_tmp_model;
+  auto x = onnxruntime::Model::Load(model_proto_sparse, p_tmp_model, nullptr, *logger_);
+
+  auto& graph2 = p_tmp_model->MainGraph();
+  auto status = graph2.Resolve();
+  EXPECT_TRUE(status.IsOK()) << status.ErrorMessage();
+  // Sparse initializer got converted to dense and appears on the list of initializers
+  ASSERT_EQ(graph2.GetAllInitializedTensors().size(), 1U);
+  ASSERT_EQ(graph2.GetAllInitializedTensors().cbegin()->first.compare(input_initializer_name), 0);
+
+  auto& graph_proto = graph2.ToGraphProto();
+  // Got propagated to initializers list
+  ASSERT_EQ(graph_proto.initializer_size(), 1);
+  ASSERT_EQ(graph_proto.initializer().at(0).name().compare(input_initializer_name), 0);
+  // Got removed from sparse initializer list
+  ASSERT_EQ(graph_proto.sparse_initializer_size(), 0);
+
+  {
+    // Check that Model::ToProto() does not return sparse among the normal initializers
+    // but reconstitutes sparse initializer from dense. Thus, here we have dense initializer list empty
+    // but it appears to be in the sparse.
+    auto model_proto_get = p_tmp_model->ToProto();
+    ASSERT_EQ(model_proto_get.graph().initializer_size(), 0);
+    ASSERT_EQ(model_proto_get.graph().sparse_initializer_size(), 1);
+  }
```

Comment:
Should we validate the contents of the sparse tensor to check it matches the original to make sure the sparse -> dense -> sparse roundtrip didn't break anything?